### PR TITLE
Support the immediate values for `u32` bitwise instructions

### PR DESCRIPTION
**Pull request by Overcastan**
This PR adds support for the immediate values for `u32` bitwise operations: `u32and`, `u32or`, `u32xor` and `u32not`.

TODO:
- [x] Update docs
- [x] Update changelog

### CODE REVIEW

**Comment by Overcastan** [comment]
It turned out that with our lalrpop parser we can just emulate instructions with immediate values with `[Push(imm), ActualInstruction]` array, without the need to create new `*Imm()` versions of instructions (of course only in case we don't need to do some other preparations and calculations). The same approach could be allied to the `u32` and `Felt` comparison instructions, so probably I should create an issue to rework them.

File: assembly/src/parser/token.rs
Comment (bobbinth) [inline assembly/src/parser/token.rs:70]: We can probably remove the comment for `Word`. Also, do we need the `Felt` variant?

File: assembly/src/parser/lexer.rs
Comment (bobbinth) [inline assembly/src/parser/lexer.rs:690]: Can we limit this to `u32` values? That is, this would become something like:
```rust
fn shrink_u32_bin(n: u32) -> BinEncodedValue
```

**Review by bobbinth — COMMENTED** [review]
Looks good! Thank you! I left a couple comments inline (both related to using only `u32` values), but also didn't review `.lalrpop` file in detail. @bitwalker could you take a look at it?

Also, we should update the changelog and documentation.

File: assembly/src/parser/token.rs
Comment (bobbinth) [reply assembly/src/parser/token.rs:70]: I would remove `Felt` for now as well. If we need it in the future, we can add it then.

File: assembly/src/parser/lexer.rs
Comment (bitwalker) [inline assembly/src/parser/lexer.rs:540]: This should be `!is_ascii_binary(c1)` no?

File: assembly/src/parser/lexer.rs
Comment (bitwalker) [inline assembly/src/parser/lexer.rs:658]: **Suggestion:** Mark this `#[inline(always)]` to ensure that it gets inlined, otherwise it goes from a very cheap comparison operation to a quite expensive function call if it isn't inlined. While Rust may inline functions like this automatically, it is based on a heuristic, and is sensitive to what is happening in the caller, so in cases where you know it is important, it's always worth providing the inline annotation

**Review by bitwalker — CHANGES_REQUESTED** [review]
Looks good overall, but I did find one bug in the lexer. Other than that just one suggestion, your call on that one.

Marking this as needing changes, but only due to the lexer bug. Just ping me for review again once its ready and I'll approve ASAP.

**Review by bitwalker — APPROVED** [review]
LGTM

File: assembly/src/parser/lexer.rs
Comment (bobbinth) [inline assembly/src/parser/lexer.rs:646]: Is this the right error type here? It's not really "Felt overflow", right?

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you! I left one comment inline, also I think https://github.com/0xPolygonMiden/miden-vm/pull/1362#discussion_r1662820708 still needs to be addressed. After these, we can merge.

File: assembly/src/parser/lexer.rs
Comment (Overcastan) [reply assembly/src/parser/lexer.rs:646]: Ah, Indeed! I overlooked it, it should be `U32Overflow`.

**Pull request merged**
